apps/api-extractor/src/generators/ExcerptBuilder.ts

```diff
@@ -214,54 +203,105 @@ export class ExcerptBuilder {
-        excerptTokens.length > 1 &&
-        !state.disableMergingForNextToken
-      ) {
-        // If the previous two tokens were a Reference and a '.', then concatenate
-        // all three tokens as a qualified name Reference.
-        const previousTokenM1: IExcerptToken = excerptTokens[excerptTokens.length - 1];
-        const previousTokenM2: IExcerptToken = excerptTokens[excerptTokens.length - 2];
+    const excerptToken: IExcerptToken = { kind: excerptTokenKind, text: text };
+    if (canonicalReference !== undefined) {
+      excerptToken.canonicalReference = canonicalReference.toString();
+    }
+    excerptTokens.push(excerptToken);
+  }
+
+  /**
+   * Condenses the provided excerpt tokens by merging tokens where possible. Updates the provided token ranges to
+   * remain accurate after token merging.
+   *
+   * @remarks
+   * For example, suppose we have excerpt tokens ["A", "B", "C"] and a token range [0, 2]. If the excerpt tokens
+   * are condensed to ["AB", "C"], then the token range would be updated to [0, 1]. Note that merges are only
+   * performed if they are compatible with the provided token ranges. In the example above, if our token range was
+   * originally [0, 1], we would not be able to merge tokens "A" and "B".
+   */
+  private static _condenseTokens(excerptTokens: IExcerptToken[], tokenRanges: IExcerptTokenRange[]): void {
+    // This set is used to quickly lookup a start or end index.
+    const startOrEndIndices: Set<number> = new Set();
+    for (const tokenRange of tokenRanges) {
+      startOrEndIndices.add(tokenRange.startIndex);
+      startOrEndIndices.add(tokenRange.endIndex);
+    }
+
+    for (let currentIndex: number = 1; currentIndex < excerptTokens.length; ++currentIndex) {
+      while (currentIndex < excerptTokens.length) {
+        const prevPrevToken: IExcerptToken = excerptTokens[currentIndex - 2]; // May be undefined
+        const prevToken: IExcerptToken = excerptTokens[currentIndex - 1];
+        const currentToken: IExcerptToken = excerptTokens[currentIndex];
+
+        // The number of excerpt tokens that are merged in this iteration. We need this to determine
+        // how to update the start and end indices of our token ranges.
+        let mergeCount: number;
+
+        // There are two types of merges that can occur. We only perform these merges if they are
+        // compatible with all of our token ranges.
         if (
-          previousTokenM1.kind === ExcerptTokenKind.Content &&
-          previousTokenM1.text.trim() === '.' &&
-          previousTokenM2.kind === ExcerptTokenKind.Reference
+          prevPrevToken &&
+          prevPrevToken.kind === ExcerptTokenKind.Reference &&
+          prevToken.kind === ExcerptTokenKind.Content &&
+          prevToken.text.trim() === '.' &&
+          currentToken.kind === ExcerptTokenKind.Reference &&
+          !startOrEndIndices.has(currentIndex) &&
+          !startOrEndIndices.has(currentIndex - 1)
```

Comment:
This "qualified name merging was copied over", it's not new code, but I think we should revisit it (in a subsequent PR). There are a few different scenarios where this merging can occur:

```
export namespace A {
  export class B {}
}

export enum C {
  D = 'D',
}

export class E {
  static f: number;
}

export function example1(): A.B {}
export function example2(): C.D {}
export function example3(): typeof E.f {}
```

There might be more examples as well. But why even merge in these cases? Isn't it more useful to have links to both `A` and `B` from `example1`, not only `B`? @octogonz maybe you can provide some historical context.

Reply:
I wrote that code on the scramble up the mountain of PR #951. Not sure how much thought was put into it. 😅